transport_analysis/viscosity.py

```diff
@@ -0,0 +1,205 @@
+    results.visc_by_particle : :class:`numpy.ndarray`
+        The viscosity of each individual particle with respect to lag-time.
+    start : Optional[int]
+        The first frame of the trajectory used to compute the analysis.
+    stop : Optional[int]
+        The frame to stop at for the analysis.
+    step : Optional[int]
+        Number of frames to skip between each analyzed frame.
+    n_frames : int
+        Number of frames analysed in the trajectory.
+    n_particles : int
+        Number of particles viscosity was calculated over.
+    """
+
+    def __init__(
+        self,
+        atomgroup: "AtomGroup",
+        temp_avg: float = 300.0,
+        dim_type: str = "xyz",
+        **kwargs,
+    ) -> None:
+        # the below line must be kept to initialize the AnalysisBase class!
+        super().__init__(atomgroup.universe.trajectory, **kwargs)
+        # after this you will be able to access `self.results`
+        # `self.results` is a dictionary-like object
+        # that can should used to store and retrieve results
+        # See more at the MDAnalysis documentation:
+        # https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html?highlight=results#MDAnalysis.analysis.base.Results
+
+        if isinstance(atomgroup, UpdatingAtomGroup):
+            raise TypeError(
+                "UpdatingAtomGroups are not valid for viscosity computation"
+            )
+
+        # args
+        self.temp_avg = temp_avg
+        self.dim_type = dim_type.lower()
+        self._dim, self.dim_fac = self._parse_dim_type(self.dim_type)
+        # self.fft = fft # consider whether fft is possible later
+
+        # local
+        self.atomgroup = atomgroup
+        self.n_particles = len(self.atomgroup)
+
+    def _prepare(self):
+        """
+        Set up viscosity, mass, velocity, and position arrays
+        before the analysis loop begins
+        """
+        # 2D viscosity array of frames x particles
+        self.results.visc_by_particle = np.zeros(
+            (self.n_frames, self.n_particles)
+        )
+
+        self._volumes = np.zeros((self.n_frames))
+
+        self._masses = self.atomgroup.masses
+        self._masses_rs = self._masses.reshape((1, len(self._masses), 1))
+
+        # 3D arrays of frames x particles x dimensionality
+        # for velocities and positions
+        self._velocities = np.zeros(
+            (self.n_frames, self.n_particles, self.dim_fac)
+        )
+
+        self._positions = np.zeros(
+            (self.n_frames, self.n_particles, self.dim_fac)
+        )
+        # self.results.timeseries not set here
+
+    @staticmethod
+    def _parse_dim_type(dim_str):
+        r"""Sets up the desired dimensionality of the calculation."""
+        keys = {
+            "x": [0],
+            "y": [1],
+            "z": [2],
+            "xy": [0, 1],
+            "xz": [0, 2],
+            "yz": [1, 2],
+            "xyz": [0, 1, 2],
```

Comment:
didn't we decide to make the keys sets? e.g. `set(["x", "y", "z"])`

Reply:
Yeah, but I'd prefer to address it in a separate PR once we're done with these ones because the current PRs are getting a bit too bulky for my tastes.